lib/mysql2/client.rb

```diff
@@ -36,7 +36,7 @@ def initialize(opts = {})
         when :reconnect, :local_infile, :secure_auth
           send(:"#{key}=", !!opts[key]) # rubocop:disable Style/DoubleNegation
         when :connect_timeout, :read_timeout, :write_timeout
-          send(:"#{key}=", opts[key].to_i)
+          send(:"#{key}=", opts[key].nil? ? nil : opts[key].to_i)
```

Comment:
Aha – it's due to accepting String values from URI params without explicit conversion. #313